### ISSUE TITLE
Plotly fails when using "scattergl" and value greater than 2^24 [Only on M1]

### ISSUE THREAD

**Issue by ravikirankhareedi** [plotly/plotly.py#3611]
When plotting using Plotly.newplot() and type scattergl on M1 machine, the plotting fails if the value of x-axis is greater than 2^24. 
Example: 

```
var data = [
  {
    x: [16777217],
    y: [1],
    type: "scattergl",
  }
];

Plotly.newPlot('myDiv', data);
```
![Screen Shot 2022-03-01 at 9 27 12 PM](https://user-images.githubusercontent.com/7555400/156300999-383d4502-030b-42bc-8da3-1c47e7fe5837.png)

 The data gets plotted on the margin which cannot be zoomed/panned. (16777217 is 2^24 + 1)
Attached screen shot. The issue only happens on M1 machine. On Intel hardware everything works well.

**Comment by c-lyu** [comment]
The same issue here. 
This issue happens only when using the `WebGL` mode. The plot becomes normal when manually switched back to the `SVG` mode.

<img width="1255" alt="image" src="https://user-images.githubusercontent.com/68106447/177357399-6f2238a2-a750-4043-9030-2cedcc76fa37.png">
<img width="1260" alt="image" src="https://user-images.githubusercontent.com/68106447/177357700-186e05cb-b0ec-458b-9fd6-c902892eb446.png">

**Comment by grorg** [comment]
This reproduces in both Chrome and Safari on M1 devices, despite them using different WebGL backends. Could plotly be sending different WebGL commands based on the architecture?

**Comment by grorg** [comment]
Or is it regl?

**Comment by djg** [comment]
Safari's implementation of WebGL passes the `-ffast-math` option to the Metal compiler and this is causing the vertex shader to be rewritten from (just concentrating on the x coord of the vertex position):

```
position.x =
        (((_ux + _utranslate.x) * _uscale.x) +
        ((_uxFract + _utranslateFract.x) * _uscale.x) +
        ((_ux + _utranslate.x) * _uscaleFract.x) +
        ((_uxFract + _utranslateFract.x) * _uscaleFract.x)) * 2.0 - 1.0f;
```
to:

```
position.x = (2.0 * (_uscaleFract.x + _uscale.x)) * ((_utranslateFract.x + _utranslate.x) + (_uxFract + _ux)) - 1.0
```

This change in the order of floating-point operations leads to a difference in the calculation causing an incorrect position:
```
position.x = (2.0 * (_uscaleFract.x + _uscale.x)) * ((_utranslateFract.x + _utranslate.x) + (_uxFract + _ux)) - 1.0
= (2.0 * (0.0 + 0.5)) * ((0.0 + -16777216.0) + (1.0 + 1677216.0)) - 1.0
```
The position and translation `16777216.0` are exactly representable because they’re power-of-2 (2 ** 24), but a 32-bit float can’t represent `16777217.0`, so `16777216.0 + 1.0` becomes `16777216.0`.
```
= (2.0 * 0.5) * (-16777216.0 + 16777216) - 1.0
= 1.0 * 0.0 - 1.0
= -1.0
```
Which is the incorrect position shown in the initial bug report.

As a workaround, `-ffast-math` can be disabled by marking `gl_Position` as `invariant` in the GLSL shader, but it should be noted that this comes with a potential impact on the performance of the shader.

**Comment by timon-schmelzer-gcx** [comment]
I have the same issue for both, Safari + Chrome and M1 Pro chip. One should note that this is also a problem when plotting a time series on the x axis, as the timestamps are internally represented as large numbers (I guess unix timestamps?).

**Comment by rickywesker** [comment]
Same issue here for M1 Pro chip + Chrome. Everything's fine using devices without the new M1 chip.
**M1**
![img_v2_c9e5de32-8fda-425f-bb84-dd670cf1c3bg](https://user-images.githubusercontent.com/20077033/211135742-6d2052b8-4b3c-419b-bc16-4783e8292682.jpg)
**Other devices**
![image](https://user-images.githubusercontent.com/20077033/211136392-f50eb14e-989a-4e88-99d9-8183b9508c73.png)

**Comment by grorg** [comment]
Yeah, @djg pointed out what the problem is, and why it only affects M1 devices. (well technically it could impact any device, but luckily/unluckily doesn't)

**Comment by alexcjohnson** [comment]
> As a workaround, `-ffast-math` can be disabled by marking `gl_Position` as `invariant` in the GLSL shader, but it should be noted that this comes with a potential impact on the performance of the shader.

@djg thanks for the tip! I'm not too worried about the performance implications, in most practical cases we're limited more by data transfer than by the shaders for these charts, but of course this remains to be tested. I'd like to understand the extent of this a bit better though before we move on to implementation, and I don't have access to an M1 device so would be grateful if others could chime in:
- Does it apply to the 3D trace types as well, if you give them the same datetime data as 2D?
- Can we replicate the problem with numeric axes as well? ie reduced precision if you have data with small variations around a large baseline? Looks like on M1 we drop down to ~10^7 (24 bits) resolution on the absolute data. Here's a codepen you can use to test: https://codepen.io/alexcjohnson/pen/RwJQGdZ - on my machine with a baseline of 10^9 I have no problem with integer resolution, but if I increase the baseline to 10^17 I see the granularity, and oddly it's different for markers and lines:
<img width="1474" alt="Screenshot 2023-01-11 at 13 27 22" src="https://user-images.githubusercontent.com/2678795/211888129-769e84e4-8efd-498a-90a2-3b61175afea6.png">

**Comment by justinjhendrick** [comment]
> Does it apply to the 3D trace types as well?

I'm seeing precision issues with 3D traces, but not specific to M1 macs. I couldn't find a difference in behavior between M1 Mac and Intel Windows 10. At the 1e9 baseline, both windows and mac are falling down.

Here's Intel Windows 10 (v22H2) on Firefox (v120.0.1):


<img width="1726" alt="Screenshot 2023-12-07 at 12 15 10 PM" src="https://github.com/plotly/plotly.py/assets/4964444/87558071-e451-4f4c-8535-0ddb25733616">


And Here's M1 mac (v13.4.1) with Firefox(v120.0.1):

![Screenshot 2023-12-07 at 12 15 44 PM](https://github.com/plotly/plotly.py/assets/4964444/48620fe6-52b0-4b63-a8f5-23e33a6846ad)

**Comment by justinjhendrick** [comment]
> As a workaround, `-ffast-math` can be disabled by marking `gl_Position` as `invariant` in the GLSL shader, but it should be noted that this comes with a potential impact on the performance of the shader.

I can confirm that @djg's suggestion fixed the issue for me. I added  `invariant gl_Position;` after these two lines in these shaders: [regl-scatter2d/marker-vert.glsl](https://github.com/gl-vis/regl-scatter2d/blob/9dbac40de176eda0daed8afb6bbed7f4ce577934/marker-vert.glsl#L6) and [regl-scatter2d/marker-vert.glsl](https://github.com/gl-vis/regl-scatter2d/blob/9dbac40de176eda0daed8afb6bbed7f4ce577934/circle-vert.glsl#L6). With those changed, `scattergl` was plotting points in the correct places with x values near 1e9.
